engine/src/main/java/org/terasology/identity/storageServiceClient/StorageServiceWorker.java

```diff
@@ -0,0 +1,183 @@
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.config.Config;
+import org.terasology.config.IdentityStorageServiceConfig;
+import org.terasology.config.SecurityConfig;
+import org.terasology.context.Context;
+import org.terasology.i18n.TranslationSystem;
+import org.terasology.identity.ClientIdentity;
+import org.terasology.identity.PublicIdentityCertificate;
+import org.terasology.logic.console.Console;
+import org.terasology.logic.console.CoreMessageType;
+
+import java.net.URL;
+import java.util.Stack;
+import java.util.Map;
+
+/**
+ * The public interface to this package. Manages a communication session with the storage service server,
+ * can answer for status information queries and can perform asynchronous operations on the storage service.
+ * This class can be in two states, with the default being "logged out".
+ */
+public final class StorageServiceWorker {
+
+    private static final Logger logger = LoggerFactory.getLogger(StorageServiceWorker.class);
+
+    StorageServiceWorkerStatus status = StorageServiceWorkerStatus.LOGGED_OUT;
+    APISession sessionInstance;
+    String loginName;
+    Stack<IdentityBundle> conflictingRemoteIdentities;
+
+    final IdentityStorageServiceConfig storageConfig;
+    final SecurityConfig securityConfig;
+
+    private final Console console;
+    private final Config config;
+    private final TranslationSystem translationSystem;
+    private Map<PublicIdentityCertificate, ClientIdentity> conflictSolutionsToUpload;
+
+    public StorageServiceWorker(Context context) {
+        this.console = context.get(Console.class);
+        this.config = context.get(Config.class);
+        this.storageConfig = this.config.getIdentityStorageService();
+        this.securityConfig = this.config.getSecurity();
+        this.translationSystem = context.get(TranslationSystem.class);
+        this.conflictingRemoteIdentities = new Stack<>();
+    }
+
+    void resetConflicts() {
+        this.conflictingRemoteIdentities = new Stack<>();
+        this.conflictSolutionsToUpload = Maps.newHashMap();
+    }
+
+    void saveConfig() {
+        config.save();
+    }
+
+    void logMessage(boolean warning, String messageId, Object... args) {
+        String localizedMessage = "Identity storage service: " + translationSystem.translate(messageId);
+        console.addMessage(String.format(localizedMessage, args), CoreMessageType.NOTIFICATION);
+    }
+
+    private synchronized void performAction(Action action, StorageServiceWorkerStatus requiredStatus) {
+        if (status != requiredStatus) {
+            throw new RuntimeException("StorageServiceWorker is not in the required status");
+        }
+        status = StorageServiceWorkerStatus.WORKING;
+        logger.info("Performing action {}", action.getClass().getSimpleName());
+        new Thread(() -> {
```

Comment:
Please provide a Runnable to the Thread instead of subclassing it. I hope we have a shared thread pool executor soon, so you can just throw it there.